core/domain/stats_services_test.py

```diff
@@ -1121,3 +1123,304 @@ def test_only_sample_answers_in_main_shard_returned(self):
+            self.assertEqual(common_elements_visualization['data'], [{
+                'answer': 'A',
+                'frequency': 6
+            }, {
+                'answer': 'B',
+                'frequency': 3
+            }, {
+                'answer': 'C',
+                'frequency': 1
+            }])
+
+    def test_retrieves_latest_vis_info_with_rounds_of_calculations(self):
+        with self._get_swap_context():
+            self._record_answer('Answer A')
+            self._record_answer('Answer C')
+            self._run_answer_summaries_aggregator()
+
+            # Submit a new answer and run the aggregator again.
+            self._record_answer('Answer A')
+            self._rerun_answer_summaries_aggregator()
+            visualizations = self._get_visualizations()
+            self.assertEqual(len(visualizations), 1)
+
+            visualization = visualizations[0]
+            # The latest data should include all submitted answers.
+            self.assertEqual(visualization['data'], [{
+                'answer': 'Answer A',
+                'frequency': 2
+            }, {
+                'answer': 'Answer C',
+                'frequency': 1
+            }])
+
+    def test_retrieves_vis_info_across_multiple_exploration_versions(self):
+        with self._get_swap_context():
+            self._record_answer('Answer A')
+            self._record_answer('Answer B')
+
+            # Change the exploration version and submit a new answer.
+            self._change_state_content('New content')
+            self._record_answer('Answer A')
+
+            self._run_answer_summaries_aggregator()
+            visualizations = self._get_visualizations()
+            self.assertEqual(len(visualizations), 1)
+
+            visualization = visualizations[0]
+            # The latest data should include all submitted answers.
+            self.assertEqual(visualization['data'], [{
+                'answer': 'Answer A',
+                'frequency': 2
+            }, {
+                'answer': 'Answer B',
+                'frequency': 1
+            }])
+
+    def test_no_vis_info_for_exp_with_new_state_name_before_calculations(self):
+        with self._get_swap_context():
+            self._record_answer('Answer A')
+            self._rename_state('new state')
+            self._run_answer_summaries_aggregator()
+
+            visualizations = self._get_visualizations(state_name='new state')
+
+            self.assertEqual(visualizations, [])
+
+    def test_no_vis_info_for_exp_with_new_state_name_after_calculations(self):
+        with self._get_swap_context():
+            self._record_answer('Answer A')
+            self._run_answer_summaries_aggregator()
+            self._rename_state('new state')
+
+            visualizations = self._get_visualizations(state_name='new state')
+
+            self.assertEqual(visualizations, [])
+
+    # TODO(bhenning): Fix these tests. These fail because we're accessing stale
```

Comment:
I don't think I understand what's going on here -- what do you mean by "fix"?

Reply:
It was fixed. :) Comment removed.